C7GameData/IProducable.cs

```diff
@@ -0,0 +1,13 @@
+namespace C7GameData
+{
+	/**
+	 * Represents something that can be produced by a city.
+	 * Known examples are Buildings and UnitPrototypes.
+	 */
+	public abstract class IProducable
```

Comment:
I'd prefer to declare these kinds of things as Interfaces with Properties, which are less brittle and better encapsulated. *deflects blame to Jon Skeet* https://csharpindepth.com/articles/PropertiesMatter

Reply:
also it seems to be spelled Producible

Reply:
As the (misspelled!) name suggests, I initially had this as an interface.  But I'm having trouble getting this interface-with-properties thing working.  I can set it up like this:

```
	public interface IProducable
	{
		string name { get; set; }
		int shieldCost { get; set; }
		int populationCost { get; set; }
	}
```

But then any attempt to access name/shieldCost/etc. fails.  E.g. `return $"{name} ({attack}/{defense}/{movement})";` says "Cannot resolve symbol 'name'".

So I'd switched to abstract class so I could access the properties from sub-classes.

I have read on Microsoft documentation that, "Beginning with C# 8.0, an interface may define a default implementation for members, including properties."  So maybe we're on too old of a C# version?

Reply:
(I'm also the opposite of a Property expert; they're one of the few things in C# that doesn't have a fairly direct equivalent in Java, so the ins and outs of when/where they can be used aren't intuitive to me yet)

Reply:
Okay, I think I figured out how to do it via StackOverflow: https://stackoverflow.com/questions/1593413/how-to-implement-a-property-in-an-interface

It seems kind of weird that it's required to put the property on the interface but also the implementing class.  Then again, I guess that was true with Java before Java 9 (which added default implementations on interfaces), too.

I see why Jon argues for properties, especially the practical benefits.  But I guess I'm still a bit confused as to why we prefer interfaces to abstract classes here.  It sort of makes sense, as in "a unit fits the interface of something that's producible".  But the abstract class phrasing, "a unit is a producible thing", also seems to fit.  In lieu of being able to think of another parent class for UnitPrototype or Building, I'm half inclined to favor the easier syntax of the abstract class over the flexibility of the interface.

Abstract classes versus interfaces are, admittedly, something that I've found myself swapping both ways in projects such as my Civ editor.  Sometimes I start with an interface and decide not having to duplicate functionality would be really nice and the "is a" metaphor fits well enough to go abstract class.  Sometimes I start abstract class and realize the metaphor isn't as solid as I'd hoped.

Anyway, I've switched it back to an interface for now (commit incoming).  Still curious about the C# 8/code duplication of properties questions though.

Reply:
Interesting, I knew about default implementations but have not worked with them. We should be on C# 8, otherwise I don't think your snippet above would even compile with fields in an interface.

Conceptually interface makes sense here because Producible describes something that it does or has via public attributes, rather than an inheritable implementation. Since C# doesn't do multiple inheritance and it's gross anyway, IMO it's best to avoid inheritance (via interfaces or composition as appropriate) until you really need to. I imagine there are other things that some Producibles would rather extend later, even though as you said those are probably Producible themselves.

Reply:
Oh and also, don't worry about it right now since we probably need to reformat everything once we figure out coding standards, but properties are always supposed to be capitalized like methods.

Reply:
Yeah, I think that makes sense.  There is no IProducible that isn't also something else, so it's an interface.

Yeah, we've kind of all been bringing out own patterns to the project.  I started a thread to try to standardize some things in December, but that's about when my enthusiasm waned, so we didn't decide on much.  This is definitely an artifact of the Java standards being different than the C# ones, and my muscle memory still being in Java.